cpp/src/arrow/buffer.h

```diff
@@ -447,14 +454,20 @@ class ARROW_EXPORT ResizableBuffer : public MutableBuffer {
 
 /// \brief Allocate a fixed size mutable buffer from a memory pool, zero its padding.
 ///
-/// \param[in] pool a memory pool
 /// \param[in] size size of buffer to allocate
-/// \param[out] out the allocated buffer (contains padding)
-///
-/// \return Status message
+/// \param[in] pool a memory pool
+ARROW_EXPORT
+Result<std::shared_ptr<Buffer>> AllocateBuffer(const int64_t size,
+                                               MemoryPool* pool = NULLPTR);
```

Comment:
```suggestion
Result<std::unique_ptr<Buffer>> AllocateBuffer(const int64_t size,
                                               MemoryPool* pool = NULLPTR);
```

Since unique_ptr is convertible to shared_ptr, this will still allow
```c++
ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> buf, AllocateBuffer(size));
```
while maintaining uniqueness of the returned buffer for anyone who doesn't wish to promote to shared_ptr.

You can then deprecate the unique_ptr Status return overload as well.

Reply:
I'll give it a try.

Reply:
Looks like GCC 7 is not able to handle the conversion `unique_ptr<ResizableBuffer> -> Result<shared_ptr<Buffer>>` https://github.com/apache/arrow/pull/6863/checks?check_run_id=568322732#step:6:663

Wrapping such conversions in an explicit `std::move` should fix it